torchrl/collectors/distributed/ray_collector.py

```diff
@@ -0,0 +1,321 @@
+
+    def _sync_iterator(self) -> Iterator[TensorDictBase]:
+
+        while self.collected_frames < self.total_frames:
+
+            # Broadcast agent weights
+            policy_weights_local_collector = {"policy_state_dict": self.local_policy.state_dict()}
+            policy_weights_local_collector_ref = ray.put(policy_weights_local_collector)
+            for e in self.remote_collectors():
+                e.load_state_dict.remote(**{"state_dict": policy_weights_local_collector_ref, "strict": False})
+
+            # Ask for batches to all remote workers.
+            pending_tasks = [e._iterator_step.remote() for e in self.remote_collectors()]
+
+            # Wait for all rollouts
+            samples_ready = []
+            while len(samples_ready) < self.num_collectors:
+                samples_ready, samples_not_ready = ray.wait(
+                    object_refs=pending_tasks, num_returns=len(pending_tasks))
+
+            # Retrieve and concatenate Tensordicts
+            out_td = []
+            for r in pending_tasks:
+                rollouts = ray.get(r)
+                ray.internal.free(r)  # should not be necessary, deleted automatically when ref count is down to 0
+                out_td.append(rollouts)
+            if len(rollouts.batch_size):
+                out_td = torch.stack(out_td)
+            else:
+                out_td = torch.cat(out_td)
+
+            self.collected_frames += out_td.numel()
+
+            yield out_td
+
+        self.stop_remote_collectors()
+        ray.shutdown()
+
+    def _async_iterator(self) -> Iterator[TensorDictBase]:
+
+        pending_tasks = {}
+        for w in self.remote_collectors():
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+        while self.collected_frames < self.total_frames:
+
+            if not len(list(pending_tasks.keys())) == len(self.remote_collectors()):
+                raise RuntimeError("Missing pending tasks, something went wrong")
+
+            # Wait for first worker to finish
+            wait_results = ray.wait(object_refs=list(pending_tasks.keys()))
+            future = wait_results[0][0]
+            w = pending_tasks.pop(future)
+
+            # Retrieve single rollouts
+            out_td = ray.get(future)
+            ray.internal.free(future)  # should not be necessary, deleted automatically when ref count is down to 0
+            self.collected_frames += out_td.numel()
+
+            # Update agent weights
+            policy_weights_local_collector = {"policy_state_dict": self.local_policy.state_dict()}
+            policy_weights_local_collector_ref = ray.put(policy_weights_local_collector)
+            w.load_state_dict.remote(**{"state_dict": policy_weights_local_collector_ref, "strict": False})
+
+            # Schedule a new collection task
+            future = w._iterator_step.remote()
+            pending_tasks[future] = w
+
+            yield out_td
```

Comment:
this should be moved before the update of the policy weights no?
Like this first the user receives the data, then the weights are updated and things keep on running

Reply:
makes sense